### PR TITLE
Restore Questasim (vsim) simulations for cv32e40s (take 2)

### PR DESCRIPTION
**Pull request by MikeOpenHWGroup**
Updates to restore Questasim of the E40S.

### CODE REVIEW

File: lib/uvm_agents/uvma_rvfi/uvma_rvfi_instr_if.sv
Comment (MikeOpenHWGroup) [inline lib/uvm_agents/uvma_rvfi/uvma_rvfi_instr_if.sv:77]: Reinstate the upstream version of this IF.

Comment (MikeOpenHWGroup) [reply lib/uvm_agents/uvma_rvfi/uvma_rvfi_instr_if.sv]: Done

File: cv32e40s/tests/uvmt/base-tests/uvmt_cv32e40s_base_test_elaboration_workarounds.sv
Comment (MikeOpenHWGroup) [reply cv32e40s/tests/uvmt/base-tests/uvmt_cv32e40s_base_test_elaboration_workarounds.sv:27]: Done

**Comment by MikeOpenHWGroup** [comment]
Hi @silabs-robin, I pushed in a couple of updates as per our discussion today.  Local CI with vsim passing.

File: cv32e40s/env/corev-dv/cv32e40s_pma_cfg.sv
Comment (silabs-robin) [reply cv32e40s/env/corev-dv/cv32e40s_pma_cfg.sv:1]: `ci_check` isn't fully green using `xrun`.
But if the PR is re-targeted, that problem might be something we can postpone til the future.

Comment (silabs-robin) [reply cv32e40s/env/corev-dv/cv32e40s_pma_cfg.sv:1]: For future reference:
bilde

File: cv32e40s/tb/uvmt/uvmt_cv32e40s_xsecure_assert/uvmt_cv32e40s_xsecure_data_independent_timing_assert.sv
Comment (silabs-krdosvik) [inline cv32e40s/tb/uvmt/uvmt_cv32e40s_xsecure_assert/uvmt_cv32e40s_xsecure_data_independent_timing_assert.sv:78]: This is okey for me. 
Just curios, were these xsecure assertions particularly problematic? I would be interested to hear why.

File: lib/isa_decoder/isa_decoder.sv
Comment (silabs-krdosvik) [reply lib/isa_decoder/isa_decoder.sv:155]: I remember there was a problem of the naming of the compressed GPRs.
I might be mistaken but I think this code was introduced to fix that naming issue (?). So to be sure this is a safe change maybe we should wait for @silabs-hfegran s review.

File: cv32e40s/env/corev-dv/cv32e40s_pma_cfg.sv
Comment (silabs-robin) [reply cv32e40s/env/corev-dv/cv32e40s_pma_cfg.sv:1]: > The PR should be re-targeted to cv32e40s/release.

Preferably, `cv32e40s/dev` should be merged to `cv32e40s/release` first (so this Questa-PR will have a proper diff after re-target).

That is likely very simple. (I expect fast-forward to be possible.)

File: lib/isa_decoder/isa_decoder.sv
Comment (silabs-robin) [reply lib/isa_decoder/isa_decoder.sv:155]: There is no need to wait. The `get_gpr_from_gpr_rvc_sreg`-changes in this PR should just be reverted. The code is there for a reason, and this PR should only be about fixing Questa, not restructuring the functionality of anything else.

File: cv32e40s/tb/uvmt/uvmt_cv32e40s_triggers_assert_cov.sv
Comment (silabs-krdosvik) [inline cv32e40s/tb/uvmt/uvmt_cv32e40s_triggers_assert_cov.sv:948]: It make sense you are wondering, for so am I.
I think this assertion is not correct. I will try to fix it.

File: cv32e40s/tb/uvmt/uvmt_cv32e40s_triggers_assert_cov.sv
Comment (silabs-krdosvik) [inline cv32e40s/tb/uvmt/uvmt_cv32e40s_triggers_assert_cov.sv:455]: Oh, interesting. how many triggers have you tested with used? 
When checking it with formal and DBG_NUM_TRIGGERS=4 it seems to work.
But I see I have not included the exception in the property (given that the is_csr_instr does not include it), which is that tselect is not accessible if DBG_NUM_TRIGGERS=0.
image
Is that the case you tested?

File: cv32e40s/tb/uvmt/uvmt_cv32e40s_triggers_assert_cov.sv
Comment (silabs-krdosvik) [inline cv32e40s/tb/uvmt/uvmt_cv32e40s_triggers_assert_cov.sv:957]: The same goes for this assertion. I think it is not checking what it is supposed to. Will fix it in an upcomming PR

File: lib/isa_decoder/isa_decoder.sv
Comment (silabs-krdosvik) [reply lib/isa_decoder/isa_decoder.sv:155]: But it is actually nice that this is attempted deleted because that indicates that the code seems useless. When/if time we can look into making this more readable/understandable. Maybe even a comment is enough.

**Comment by silabs-robin** [comment]
Continued in https://github.com/openhwgroup/core-v-verif/pull/2320

**Comment by silabs-robin** [comment]
Since I don't want to transfer all of the unresolved comments, I shall see if I can "revive" this PR instead of using the new one.

File: cv32e40s/env/corev-dv/cv32e40s_pma_cfg.sv
Comment (silabs-robin) [reply cv32e40s/env/corev-dv/cv32e40s_pma_cfg.sv:1]: > > The PR should be re-targeted to cv32e40s/release.
> 
> Preferably, `cv32e40s/dev` should be merged to `cv32e40s/release` first (so this Questa-PR will have a proper diff after re-target).
> 
> That is likely very simple. (I expect fast-forward to be possible.)

Done. https://github.com/openhwgroup/core-v-verif/pull/2323

Comment (silabs-robin) [reply cv32e40s/env/corev-dv/cv32e40s_pma_cfg.sv:1]: > The PR should be re-targeted to cv32e40s/release.
>
> I know that will be inconvenient, so sorry about that.

Actually, it was not inconvenient, one can just press "edit" and change the base.

**Review by silabs-robin — COMMENTED** [review]
I still have a feedback comment open that I want resolved before I can approve of this.

File: lib/isa_decoder/isa_decoder.sv
Comment (silabs-robin) [reply lib/isa_decoder/isa_decoder.sv:155]: @MikeOpenHWGroup

File: cv32e40s/tb/uvmt/uvmt_cv32e40s_xsecure_assert/uvmt_cv32e40s_xsecure_data_independent_timing_assert.sv
Comment (silabs-robin) [reply cv32e40s/tb/uvmt/uvmt_cv32e40s_xsecure_assert/uvmt_cv32e40s_xsecure_data_independent_timing_assert.sv:78]: > Making the sequence a clocked sequence clears the compile error

Btw, when that gets fixed later, we will also be closer to Questa Formal working.

`cd core-v-verif/cv32e40s/fv` and do `make questa` if you wanna try it out.

File: lib/isa_decoder/isa_decoder.sv
Comment (MikeOpenHWGroup) [reply lib/isa_decoder/isa_decoder.sv]: I pushed in an update to restore this function.

File: lib/isa_decoder/isa_decoder.sv
Comment (MikeOpenHWGroup) [reply lib/isa_decoder/isa_decoder.sv:1537]: I pushed in an update to exclude this only for Questa.

File: cv32e40s/tb/uvmt/uvmt_cv32e40s_triggers_assert_cov.sv
Comment (silabs-krdosvik) [reply cv32e40s/tb/uvmt/uvmt_cv32e40s_triggers_assert_cov.sv:948]: I now remember why the assertions are written accordingly. (The trigger match support logic is a bit complicated so tried to avoid modifying to make it fit this scenario, so I started thinking about the alternatives).

DCSR and DPC are written if: 1) there is a csr write operation or 2) as a sideeffect of entering debug mode.
If the core is already in debug mode, but see a debug entery action such as a trigger match the DCSR and DPC are not updated (since the core is already in debug mode).

We can therefor verify that the core does no action when seeing trigger matches in debug mode by making sure the if DCSR or DPC are written to (wmask != 0) it must have been a csr write operation (and not a sideeffect of a trigger match).

Comment (silabs-krdosvik) [reply cv32e40s/tb/uvmt/uvmt_cv32e40s_triggers_assert_cov.sv:948]: But I will add a comment about it in the code to make it more clear.

File: cv32e40s/tb/uvmt/uvmt_cv32e40s_triggers_assert_cov.sv
Comment (silabs-krdosvik) [reply cv32e40s/tb/uvmt/uvmt_cv32e40s_triggers_assert_cov.sv:957]: Same ^^

File: cv32e40s/tb/uvmt/uvmt_cv32e40s_xsecure_assert/uvmt_cv32e40s_xsecure_data_independent_timing_assert.sv
Comment (silabs-krdosvik) [reply cv32e40s/tb/uvmt/uvmt_cv32e40s_xsecure_assert/uvmt_cv32e40s_xsecure_data_independent_timing_assert.sv:78]: Thanks! I think it should be okay to use clocked sequences.

File: cv32e40s/tb/uvmt/uvmt_cv32e40s_triggers_assert_cov.sv
Comment (silabs-krdosvik) [reply cv32e40s/tb/uvmt/uvmt_cv32e40s_triggers_assert_cov.sv:455]: I see. Running make test TEST=debug_test2 (and using our default simulator which I believe is xcelium) i dont see any errors.

File: lib/isa_decoder/isa_decoder.sv
Comment (silabs-robin) [reply lib/isa_decoder/isa_decoder.sv]: > I pushed in an update to restore this function.

!

**Review by silabs-robin — APPROVED** [review]
Approving based on:
Content of review discussions.
All "please fix this"-comments have been resolved.

**Pull request merged**
